Race condition in proxy store

**Thanos, Prometheus and Golang version used**: 
Thanos version: thanosio/thanos:main-2022-06-23-f60b6598

**What happened**:
A Querier crashed several times with the following message

> Message:   east1-2\", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"},{k8s_cluster="[redacted]", replica=\"thanos-receiver-17\"} Mint: -9223372036854775808 Maxt: 9223372036854775807: rpc error: code = Canceled desc = grpc: the client connection is closing"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x158f678]

> goroutine 631657 [running]:
github.com/thanos-io/thanos/pkg/query.(*endpointRef).Series(0xc000cb9a40, {0x21fe8d0, 0xc00ba115f0}, 0xc, {0x0, 0x1, 0x1})
  <autogenerated>:1 +0x38
github.com/thanos-io/thanos/pkg/store.(*ProxyStore).Series.func1()
  /app/pkg/store/proxy.go:314 +0xf5f
golang.org/x/sync/errgroup.(*Group).Go.func1()
  /go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x67
created by golang.org/x/sync/errgroup.(*Group).Go
  /go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x92


**What you expected to happen**:
I would expect Querier to handle the removal of a store gracefully while executing a query.

**How to reproduce it (as minimally and precisely as possible)**:
The issue is hard to reproduce locally since it seems to be caused by a race condition when a query is executed and a store is removed from the endpoints at the same time.

This is where the panic originates from
https://github.com/thanos-io/thanos/blob/f644d75586e60d86e2861cb5a9bceb7bee98188b/pkg/store/proxy.go#L314

--- a/pkg/query/endpointset.go
+++ b/pkg/query/endpointset.go
@@ -61,17 +61,17 @@ func (es *GRPCEndpointSpec) Addr() string {
 
 // Metadata method for gRPC endpoint tries to call InfoAPI exposed by Thanos components until context timeout. If we are unable to get metadata after
 // that time, we assume that the host is unhealthy and return error.
-func (es *GRPCEndpointSpec) Metadata(ctx context.Context, client *endpointClients) (*endpointMetadata, error) {
-	if client.info != nil {
-		resp, err := client.info.Info(ctx, &infopb.InfoRequest{}, grpc.WaitForReady(true))
+func (es *GRPCEndpointSpec) Metadata(ctx context.Context, infoClient infopb.InfoClient, storeClient storepb.StoreClient) (*endpointMetadata, error) {
+	if infoClient != nil {
+		resp, err := infoClient.Info(ctx, &infopb.InfoRequest{}, grpc.WaitForReady(true))
 		if err == nil {
 			return &endpointMetadata{resp}, nil
 		}
 	}
 
 	// Call Info method of StoreAPI, this way querier will be able to discovery old components not exposing InfoAPI.
-	if client.store != nil {
-		metadata, err := es.getMetadataUsingStoreAPI(ctx, client.store)
+	if storeClient != nil {
+		metadata, err := es.getMetadataUsingStoreAPI(ctx, storeClient)
 		if err != nil {
 			return nil, errors.Wrapf(err, "fallback fetching info from %s", es.addr)
 		}
@@ -368,7 +368,12 @@ func (e *EndpointSet) GetStoreClients() []store.Client {
 	stores := make([]store.Client, 0, len(e.endpoints))
 	for _, er := range e.endpoints {
 		if er.HasStoreAPI() {
-			stores = append(stores, er)
+			// Make a new endpointRef with store client.
+			stores = append(stores, &endpointRef{
+				StoreClient: storepb.NewStoreClient(er.cc),
+				addr:        er.addr,
+				metadata:    er.metadata,
+			})
 		}
 	}
 	return stores
@@ -382,7 +387,7 @@ func (e *EndpointSet) GetQueryAPIClients() []querypb.QueryClient {
 	stores := make([]querypb.QueryClient, 0, len(e.endpoints))
 	for _, er := range e.endpoints {
 		if er.HasQueryAPI() {
-			stores = append(stores, er.clients.query)
+			stores = append(stores, querypb.NewQueryClient(er.cc))
 		}
 	}
 	return stores
@@ -396,7 +401,7 @@ func (e *EndpointSet) GetRulesClients() []rulespb.RulesClient {
 	rules := make([]rulespb.RulesClient, 0, len(e.endpoints))
 	for _, er := range e.endpoints {
 		if er.HasRulesAPI() {
-			rules = append(rules, er.clients.rule)
+			rules = append(rules, rulespb.NewRulesClient(er.cc))
 		}
 	}
 	return rules
@@ -410,7 +415,7 @@ func (e *EndpointSet) GetTargetsClients() []targetspb.TargetsClient {
 	targets := make([]targetspb.TargetsClient, 0, len(e.endpoints))
 	for _, er := range e.endpoints {
 		if er.HasTargetsAPI() {
-			targets = append(targets, er.clients.target)
+			targets = append(targets, targetspb.NewTargetsClient(er.cc))
 		}
 	}
 	return targets
@@ -424,7 +429,7 @@ func (e *EndpointSet) GetMetricMetadataClients() []metadatapb.MetadataClient {
 	metadataClients := make([]metadatapb.MetadataClient, 0, len(e.endpoints))
 	for _, er := range e.endpoints {
 		if er.HasMetricMetadataAPI() {
-			metadataClients = append(metadataClients, er.clients.metricMetadata)
+			metadataClients = append(metadataClients, metadatapb.NewMetadataClient(er.cc))
 		}
 	}
 	return metadataClients
@@ -439,7 +444,7 @@ func (e *EndpointSet) GetExemplarsStores() []*exemplarspb.ExemplarStore {
 	for _, er := range e.endpoints {
 		if er.HasExemplarsAPI() {
 			exemplarStores = append(exemplarStores, &exemplarspb.ExemplarStore{
-				ExemplarsClient: er.clients.exemplar,
+				ExemplarsClient: exemplarspb.NewExemplarsClient(er.cc),
 				LabelSets:       labelpb.ZLabelSetsToPromLabelSets(er.metadata.LabelSets...),
 			})
 		}
@@ -498,14 +503,10 @@ func (e *EndpointSet) getActiveEndpoints(ctx context.Context, endpoints map[stri
 					cc:     conn,
 					addr:   addr,
 					logger: e.logger,
-					clients: &endpointClients{
-						info:  infopb.NewInfoClient(conn),
-						store: storepb.NewStoreClient(conn),
-					},
 				}
 			}
 
-			metadata, err := spec.Metadata(ctx, er.clients)
+			metadata, err := spec.Metadata(ctx, infopb.NewInfoClient(er.cc), storepb.NewStoreClient(er.cc))
 			if err != nil {
 				if !seenAlready && !spec.IsStrictStatic() {
 					// Close only if new and not a strict static node.
@@ -622,8 +623,6 @@ type endpointRef struct {
 	cc   *grpc.ClientConn
 	addr string
 
-	clients *endpointClients
-
 	// Metadata can change during runtime.
 	metadata *endpointMetadata
 
@@ -634,42 +633,6 @@ func (er *endpointRef) Update(metadata *endpointMetadata) {
 	er.mtx.Lock()
 	defer er.mtx.Unlock()
 
-	clients := er.clients
-
-	if metadata.Store != nil {
-		clients.store = storepb.NewStoreClient(er.cc)
-		er.StoreClient = clients.store
-	} else {
-		// When we see the endpoint for the first time we assume the StoreAPI is exposed by that endpoint (which may not be true for some component, e.g. ruler)
-		// and we create a store API client because as a fallback we might have to call info method of storeAPI.
-		// In this step, we are setting it to null when we find out that the store API is not exposed.
-		er.clients.store = nil
-		er.StoreClient = nil
-	}
-
-	if metadata.Rules != nil {
-		clients.rule = rulespb.NewRulesClient(er.cc)
-	}
-
-	if metadata.Targets != nil {
-		clients.target = targetspb.NewTargetsClient(er.cc)
-	}
-
-	if metadata.MetricMetadata != nil {
-		clients.metricMetadata = metadatapb.NewMetadataClient(er.cc)
-	}
-
-	if metadata.Exemplars != nil {
-		// min/max range is also provided by in the response of Info rpc call
-		// but we are not using this metadata anywhere right now so ignoring.
-		clients.exemplar = exemplarspb.NewExemplarsClient(er.cc)
-	}
-
-	if metadata.Query != nil {
-		clients.query = querypb.NewQueryClient(er.cc)
-	}
-
-	er.clients = clients
 	er.metadata = metadata
 }
 
@@ -688,42 +651,42 @@ func (er *endpointRef) HasStoreAPI() bool {
 	er.mtx.RLock()
 	defer er.mtx.RUnlock()
 
-	return er.clients != nil && er.clients.store != nil
+	return er.metadata != nil && er.metadata.Store != nil
 }
 
 func (er *endpointRef) HasQueryAPI() bool {
 	er.mtx.RLock()
 	defer er.mtx.RUnlock()
 
-	return er.clients != nil && er.clients.query != nil
+	return er.metadata != nil && er.metadata.Query != nil
 }
 
 func (er *endpointRef) HasRulesAPI() bool {
 	er.mtx.RLock()
 	defer er.mtx.RUnlock()
 
-	return er.clients != nil && er.clients.rule != nil
+	return er.metadata != nil && er.metadata.Rules != nil
 }
 
 func (er *endpointRef) HasTargetsAPI() bool {
 	er.mtx.RLock()
 	defer er.mtx.RUnlock()
 
-	return er.clients != nil && er.clients.target != nil
+	return er.metadata != nil && er.metadata.Targets != nil
 }
 
 func (er *endpointRef) HasMetricMetadataAPI() bool {
 	er.mtx.RLock()
 	defer er.mtx.RUnlock()
 
-	return er.clients != nil && er.clients.metricMetadata != nil
+	return er.metadata != nil && er.metadata.MetricMetadata != nil
 }
 
 func (er *endpointRef) HasExemplarsAPI() bool {
 	er.mtx.RLock()
 	defer er.mtx.RUnlock()
 
-	return er.clients != nil && er.clients.exemplar != nil
+	return er.metadata != nil && er.metadata.Exemplars != nil
 }
 
 func (er *endpointRef) LabelSets() []labels.Labels {
@@ -801,16 +764,6 @@ func (er *endpointRef) apisPresent() []string {
 	}
 
 	return apisPresent
-}
-
-type endpointClients struct {
-	store          storepb.StoreClient
-	rule           rulespb.RulesClient
-	metricMetadata metadatapb.MetadataClient
-	exemplar       exemplarspb.ExemplarsClient
-	target         targetspb.TargetsClient
-	query          querypb.QueryClient
-	info           infopb.InfoClient
 }
 
 type endpointMetadata struct {

--- a/pkg/query/endpointset_test.go
+++ b/pkg/query/endpointset_test.go
@@ -1193,7 +1193,6 @@ func TestDeadlockLocking(t *testing.T) {
 		metadata: &endpointMetadata{
 			&infopb.InfoResponse{},
 		},
-		clients: &endpointClients{},
 	}
 
 	g := &errgroup.Group{}

--- a/test/e2e/query_test.go
+++ b/test/e2e/query_test.go
@@ -1276,7 +1276,7 @@ func TestSidecarAlignmentPushdown(t *testing.T) {
 func TestGrpcInstantQuery(t *testing.T) {
 	t.Parallel()
 
-	e, err := e2e.NewDockerEnvironment("e2e_test_query_grpc_api")
+	e, err := e2e.NewDockerEnvironment("e2e_test_query_grpc_api_instant")
 	testutil.Ok(t, err)
 	t.Cleanup(e2ethanos.CleanScenario(t, e))
 
@@ -1382,7 +1382,7 @@ func TestGrpcInstantQuery(t *testing.T) {
 func TestGrpcQueryRange(t *testing.T) {
 	t.Parallel()
 
-	e, err := e2e.NewDockerEnvironment("e2e_test_query_grpc_api")
+	e, err := e2e.NewDockerEnvironment("e2e_test_query_grpc_api_range")
 	testutil.Ok(t, err)
 	t.Cleanup(e2ethanos.CleanScenario(t, e))
 

